tools/pre-push

```diff
@@ -0,0 +1,83 @@
+#!/bin/sh
+
+set -eu
+
+if [ -t 2 ]; then
+    red="\e[1;31m"
+    green="\e[1;32m"
+    blue="\e[1;34m"
+    grey="\e[0m"
+else
+    red=""
+    green=""
+    blue=""
+    grey=""
+fi
+
+# checks a single commit (ref name in $1) for validity
+check_commit() {
+    tmpdir="tmp/pre-push/${commit}"
+    rm -rf "tmp/pre-push/${commit}"
+
+    # unpack all of the changed files, plus the test-static-code from that commit
+    files="$(git show --format='' --name-only "${commit}" --) tools/test-static-code"
+    git archive --prefix="${tmpdir}/" "${commit}" -- ${files} | tar x
+
+    # check that node_modules always gets updated with package.json
+    if [ -e "${tmpdir}/package.json" -a ! -e "${tmpdir}/node_modules" ]; then
+        echo 'not ok 0 /node-modules/update'
+        echo '# package.json changed, but node_modules not updated'
+    fi
+
+    # if any of the following commands fail, the function
+    # will abort, leaving the temp directory in place
+    (
+        cd "${tmpdir}"
+        env -i git init -qb main .
+        env -i git add .
+        ! tools/test-static-code | grep '^[#n]'
+    )
+
+    # paranoia: make sure we delete what we think we will
+    rm -rf "tmp/pre-push/${commit}"
+}
+
+main() {
+    nothing='0000000000000000000000000000000000000000'
+    fail=''
+
+    # cf. man 5 githooks
+    # <local ref> SP <local object name> SP <remote ref> SP <remote object name> LF
+    while read local_ref local_object_name remote_ref remote_object_name; do
+        printf "${blue}Performing pre-push checks for %s:${grey}\n" "${remote_ref}" >&2
+
+        if [ "${local_object_name}" = "${nothing}" ]; then
+            # deleting remote ref: no checks
+            continue
+        fi
+
+        if [ "${remote_object_name}" != "${nothing}" ]; then
+            reference="${remote_object_name}"
+        else
+            reference="origin/main"
```

Comment:
Having to hardcode a branch name here feels wrong. If git can't figure it out by itself somehow, it's almost certainly going to be broken for stable branches like origin/rhel-8.5

Reply:
Another idea: Once we submit a PR, then the origin target is known. Could this "check all commits" move into a workflow instead, and as the pre-push hook we just keep the "check the current tree" behaviour that we have now? It's not like the pre-push hook is going to find all problems with your PRs anyway, it's just a tool to tell you the most glaring problems.

Reply:
This doesn't help much, though: the main benefit here is that we only check changed files, but without being able to determine what the "starting point is" we also can't determine the list of files changed by the current branch...

I remembered that `git merge-base` has a `--fork-point` option which somehow does some magic using the reflog to try to guess where you "started work".  I'll take a look at that next...

Otherwise, it would be really awesome if we had some way to be able to tell commits that were created by GitHub (during "rebase and merge") and commits that were created by human users of `git`.  I don't know any, though.

Reply:
It helps with being thorough (every commit passes individually). It of course does not help with testing only changed files locally, but then again, if you push your PRs so often that 7.5s matter, we have bigger problems with our CI capacity :wink: I can't say I'm very thrilled about starting guesswork and heuristics here..

Reply:
origin == cockpit-project/cockpit works for me, and might work for you, but ISTR that @mvollmer has a different workflow where origin == his fork. I think that does *not* invalidate your reasoning, as these pushes would also have been pre-verified. I just wanted to point it out here. It's not that important, but the one thing we need to avoid is to erroneously check hundreds of commits.